### ISSUE TITLE
Change usernames to display names

### ISSUE THREAD

**Issue by dylex** [dylex/slack-libpurple#19]
https://api.slack.com/changelog/2017-09-the-one-about-usernames
#13

**Comment by dylex** [comment]
The problem with this is that display names are not unique, and the purple API requires unique names that are also displayed to users for pretty much all interaction.  We may have to fall back to `display_name (id)` or something if it comes to that.

**Comment by lf94** [comment]
> display_name (id)

Good idea. Maybe make it the default even. I have my own fork to support showing links too. I will do this as well.

**Comment by lf94** [comment]
Not sure what you mean @EionRobb , I'm not familiar with bitlbee. I only know what I need and how to go about it. Please describe more what *you* need and I'll consider, if it seems like a better approach.

**Comment by EionRobb** [comment]
The `serv_got_alias()` function sets the server alias on a buddy, which is how other prpls would show the full name of a buddy rather than their id, while keeping the internal id available for api calls.  In Bitlbee-land this gets mapped to the `%full_name` parameter of the `nick_format` setting, eg  `account slack set nick_format %full_name`

**Comment by lf94** [comment]
Eh, I just went ahead and did this:

```
diff --git a/slack-user.c b/slack-user.c
index 50a4d4e..18744bb 100644
--- a/slack-user.c
+++ b/slack-user.c
@@ -52,8 +52,25 @@ SlackUser *slack_user_update(SlackAccount *sa, json_value *json) {
 		g_hash_table_replace(sa->users, user->object.id, user);
 	}
 
-	const char *name = json_get_prop_strptr(json, "name");
-	g_warn_if_fail(name);
+	char *username = json_get_prop_strptr(json, "name");
+	g_warn_if_fail(username);
+
+	json_value *profile = json_get_prop_type(json, "profile", object);
+	char *display_name = NULL;
+	if (profile) {
+		display_name = json_get_prop_strptr(profile, "display_name_normalized");
+		g_warn_if_fail(display_name);
+	}
+
+	gchar *name = NULL;
+	if (g_strcmp0(display_name, username) == 0
+		|| g_strcmp0(display_name, "") == 0
+		|| display_name == NULL
+	) {
+		name = username;
+	} else {
+		name = g_strconcat(display_name, "|", username, NULL);
+	}
 
 	if (g_strcmp0(user->object.name, name)) {
 		purple_debug_misc("slack", "user %s: %s\n", sid, name);
@@ -67,7 +84,6 @@ SlackUser *slack_user_update(SlackAccount *sa, json_value *json) {
 			purple_blist_rename_buddy(user_buddy(user), user->object.name);
 	}
 
-	json_value *profile = json_get_prop_type(json, "profile", object);
 	if (profile) {
 		const char *status = json_get_prop_strptr1(profile, "status_text") ?: json_get_prop_strptr1(profile, "current_status");
 		g_free(user->status);
```

Enjoy!

**Comment by EionRobb** [comment]
Wouldn't you then have to change everywhere else that the name is used to handle a name with a pipe in it, such as sending IM's?
Why not instead just call `serv_got_alias(sa->gc, username, display_name);`

**Comment by lf94** [comment]
Nope, I was able to send IMs just normally.

**Comment by dylex** [comment]
@EionRobb's suggestion sounds reasonable so I've added it for now. However, none of this addresses the real issue, that slack plans to eliminate usernames entirely, so we have to move away from all uses of the name field. The main problem with this is dealing with `@` references (which @lf94's code can break by introducing spaces), but we may have to give up on that anyway (at least in my ongoing work on lazy user loading).

(Also, @lf94, your code leaks the result of g_strconcat.)
